firmware/applications/SignBuddy/comms.c

```diff
@@ -0,0 +1,150 @@
+#include "comms.h"
+
+#include "ble_uart.h"
+#include "board.h"
+#include "common.h"
+#include "crc.h"
+#include "fifo.h"
+#include "logger.h"
+#include "pb_encode.h"
+#include "SignBuddy.pb.h"
+
+#include <stdlib.h>
+
+#define COMMS_TX_BUFFER_SIZE    512
+
+typedef struct __attribute__((__packed__)) {
+  uint8_t length;
+  uint8_t sample[Sample_size];
+  uint32_t crc;
+} packet_t;
+
+typedef struct {
+  TaskHandle_t      task_handle;
+  Sample            sample;
+  packet_t          packet;
+  volatile uint8_t  sample_ready;
+  volatile uint8_t  packet_ready;
+  uint8_t           tx_buffer[COMMS_TX_BUFFER_SIZE];
+  fifo_t            tx_fifo;
+  SemaphoreHandle_t sample_mutex;
+} state_t;
+
+static state_t s;
+
+static void hw_init(void)
+{
+  ble_uart_init();
+}
+
+static void rx()
+{
+  DISABLE_IRQ();
+  uint8_t cmd = LL_LPUART_ReceiveData8(BLE_UART);
+  ENABLE_IRQ();
+
+  switch (cmd) {
+  case 0:
+    LOG_DEBUG("comms: cmd 1\r\n");
+    break;
+
+  case 1:
+    LOG_DEBUG("comms: cmd 2\r\n");
+    break;
+
+  case 2:
+    LOG_DEBUG("comms: cmd 3\r\n");
+    break;
+
+  default:
+    LOG_DEBUG("comms: unknown cmd\r\n");
+    break;
+  }
+}
+
+static void ingest_packet()
+{
+  fifo_push(&s.tx_fifo, (uint8_t *) &s.packet, sizeof(s.packet));
```

Comment:
you shouldn't push `sizeof(s.packet)` bytes.

you should push `s.packet.length + whatever` bytes